File: neet_code_all_cpp/array_hash/brick_wall.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

// 模範解答
class SolutionAns {
    public:
        int leastBricks(std::vector<std::vector<int>>& wall) {
            std::unordered_map<int,int> end_count;
            int max_end_count = 0;
            int rows = wall.size();

            for(int i = 0; i < rows; ++i) {
                int end_of_brick = 0; 
                int cols = wall[i].size() - 1;

                for(int j = 0; j < cols; ++j) {
                    end_of_brick += wall[i][j];

                    if(end_count.find(end_of_brick) != end_count.end()) {
                        ++end_count[end_of_brick];
                    } else {
                        end_count[end_of_brick] = 1;
                    }

                    max_end_count = std::max(max_end_count, end_count[end_of_brick]);
                }
            }

            return rows - max_end_count;
        }
};

int main(void) {
    using Case = std::vector<std::vector<int>>;

    Case case_1 = {{1,2,2,1},{3,1,2},{1,3,2},{2,4},{3,1,2},{1,3,1,1}};
    // => 2
    Case case_2 = {{1},{1},{1}};
    // => 3

    SolutionAns s_ans;
    std::cout<< s_ans.leastBricks(case_1) << std::endl;
    std::cout<< s_ans.leastBricks(case_2) << std::endl;
}

File: neet_code_all/src/array_hash/brick_wall.rs
use std::collections::HashMap;

// 解けなかった
struct Solution {}
impl Solution {
    pub fn least_bricks(wall: Vec<Vec<i32>>) -> i32 {
        let mut map: HashMap<i32, i32> = HashMap::new();

        1
    }
}

// AC
// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn least_bricks(wall: Vec<Vec<i32>>) -> i32 {
        let mut end_count: HashMap<i32, i32> = HashMap::new();
        let (mut max_end_count, rows) = (0, wall.len());

        for i in 0..rows {
            let mut end_of_brick = 0;
            let cols = wall[i].len() - 1;

            for j in 0..cols {
                end_of_brick += wall[i][j];

                if let Some(end_count_key) = end_count.get_mut(&end_of_brick) {
                    *end_count_key += 1;
                } else {
                    end_count.insert(end_of_brick, 1);
                }

                max_end_count = max_end_count.max(end_count[&end_of_brick]);
            }
        }

        rows as i32 - max_end_count
    }
}

fn main() {
    let case_1 = vec![
        vec![1, 2, 2, 1],
        vec![3, 1, 2],
        vec![1, 3, 2],
        vec![2, 4],
        vec![3, 1, 2],
        vec![1, 3, 1, 1],
    ];
    // => 2
    let case_2 = vec![vec![1], vec![1], vec![1]];
    // => 3

    println!("case_1: {}", SolutionAnsCpp::least_bricks(case_1.clone()));
    println!("case_2: {}", SolutionAnsCpp::least_bricks(case_2.clone()));
}
